x-pack/plugins/aiops/public/components/log_categorization/use_categorize_request.ts

```diff
@@ -0,0 +1,196 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { cloneDeep } from 'lodash';
+import { useRef, useCallback } from 'react';
+import { isCompleteResponse, isErrorResponse } from '@kbn/data-plugin/public';
+
+import { useAiopsAppContext } from '../../hooks/use_aiops_app_context';
+
+const CATEGORY_LIMIT = 1000;
+const EXAMPLE_LIMIT = 1;
+
+interface CatResponse {
+  rawResponse: {
+    aggregations: {
+      categories: {
+        buckets: Array<{
+          key: string;
+          doc_count: number;
+          hit: { hits: { hits: Array<{ _source: { message: string } }> } };
+          sparkline: { buckets: Array<{ key_as_string: string; key: number; doc_count: number }> };
+        }>;
+      };
+    };
+  };
+}
+
+export interface Category {
+  key: string;
+  count: number;
+  examples: string[];
+  sparkline?: Array<{ doc_count: number; key: number; key_as_string: string }>;
+}
+
+export type EventRate = Array<{
+  key: number;
+  docCount: number;
+}>;
+
+export type SparkLinesPerCategory = Record<string, Record<number, number>>;
+
+export function useCategorizeRequest() {
+  const { data } = useAiopsAppContext();
+
+  const abortController = useRef(new AbortController());
+
+  const runCategorizeRequest = useCallback(
+    (
+      index: string,
+      field: string,
+      timeField: string,
+      from: number | undefined,
+      to: number | undefined,
+      queryIn: any,
+      intervalMs?: number
+    ): Promise<{ categories: Category[]; sparkLinesPerCategory: SparkLinesPerCategory }> => {
+      return new Promise((resolve, reject) => {
+        data.search
+          .search<any, CatResponse>(
```

Comment:
do we have a better type definition?

Reply:
Updated in https://github.com/elastic/kibana/pull/139005/commits/41a5137ab0cd82a44b637cdd28e15af46f9fcc46